R/share_spec.R

```diff
@@ -129,49 +144,63 @@ share_spec.data.frame <- function(data,
                                                level_of_confidence_in_identification = "",
                                                other_info = "",
                                                license = "CC BY-NC"),
-                                  share = "system", ...) {
+                                  file = NULL,
+                                  share = "system",
+                                  uid = digest(sessionInfo()), ...) {
   if (is.null(names(metadata))) stop("'metadata' needs to be a named vector")
   if (any(is.na(metadata[c("user_name", "spectrum_type", "spectrum_identity")])) |
       metadata["user_name"] == "" | metadata["spectrum_type"] == "" |
       metadata["spectrum_identity"] == "") {
     mex <- FALSE
     warning("fields 'user_name', 'spectrum_type', and 'spectrum_identity' ",
-            "must not be empty if you like to share your metadata")
+            "must not be empty if you like to share your metadata", call. = F)
   } else {
     mex <- TRUE
   }
 
-  id <- paste(human_ts(), digest(data, algo = "md5"), sep = "_")
+  fid <- paste(human_ts(), digest(data), sep = "_")
 
   mdata <- data.frame(variable = names(metadata), input = metadata,
                       row.names = NULL)
 
   if (share == "system") {
-    pkg <- system.file("extdata", package = "OpenSpecy")
-    dir.create(file.path(pkg, "user_spectra"), showWarnings = F)
-    fp <- file.path(pkg, "user_spectra")
+    fp <- file.path(system.file("extdata", package = "OpenSpecy"),
+                    "user_spectra", uid)
+    dir.create(fp, showWarnings = F)
   } else if (share == "dropbox") {
-    if (!requireNamespace("rdrop2"))
-      stop("share = 'dropbox' requires package 'rdrop2'")
-    fp <- file.path(tempdir())
+
+    pkg <- "rdrop2"
+    mpkg <- pkg[!(pkg %in% installed.packages()[ , "Package"])]
+    if (length(mpkg)) stop("share = 'dropbox' requires package 'rdrop2'")
+
+    fp <- file.path(tempdir(), uid)
+    dir.create(fp, showWarnings = F)
   } else {
-    if (!dir.exists(share)) dir.create(share, showWarnings = F)
-    fp <- file.path(share)
+    fp <- file.path(share, uid)
+    dir.create(fp, showWarnings = F)
   }
 
-  fd <- file.path(fp, paste0(id, ".csv"))
-  fm <- file.path(fp, paste0(id, "_form", ".csv"))
+  fd <- file.path(fp, paste0(fid, ".csv"))
+  fm <- file.path(fp, paste0(fid, "_form", ".csv"))
 
   write.csv(data, fd, row.names = FALSE, quote = TRUE)
 
   if (mex) write.csv(mdata, fm, row.names = FALSE, quote = TRUE)
+  if (!is.null(file)) {
+    ex <- strsplit(basename(file), split="\\.")[[1]]
+    file.copy(file, file.path(fp, paste0(fid, ".", ex[-1])))
+  }
 
   if (share == "dropbox") {
-    rdrop2::drop_upload(fd, path = "Spectra", ...)
-    if (mex) rdrop2::drop_upload(fm, path = "Spectra", ...)
+    for (lf in list.files(fp, full.names = T)) {
+      rdrop2::drop_upload(lf, path = paste0("data/users/", uid), ...)
```

Comment:
#54 We need to add a session-id after users here and above for fp?

Reply:
Ah, got it! How are those session IDs made? I'll need to find a surrogate for the local package users. Just like `uid = digest(sessionInfo())`. But maybe this isn't really necessary. For the Shiny part, we just needed a new argument (now `uid`) that can be fed to the generic sharing functions. How we use them (for a user id only or by pasting together two ideas) is up to us.

Reply:
Okay, I now use both `Sys.info()` and `sessionInfo()` just to keep everything comparable. We'll need different IDs for the Shiny interface anyway.

Reply:
That will work! Thanks. The session IDs were being made by creating a unique random identifier inside of the server function so that every time a session started a new id would be made. The user id for the shiny app is made by hashing the browser information. We can add these to the config files.

Reply:
Just saw it .. with `random_id()`

Reply:
Sys.info will return just the shiny server system on shiny.io. Session info will likely be the same for every day on shiny.io too. We should probably set the defaults differently for shiny.io and desktop. I still think the ids package should be used for desktop instead of session.info. We want the session id to be unique to whenever the user opens the app and sessionInfo wont be. Sys.info seems fine for desktop though and probably better than hashing the browser.

Reply:
Yeah, this is just some defaults for the package function. We'll use the browser ID and a randomized session ID for Shiny using the function arguments. But I'll also default to a random number for the package instead of using the sessionInfo. Anyway, does it need to be ids:random_id? We could also hash `sample()`